### ISSUE TITLE
Timepoints are not unique at triggered events

### ISSUE THREAD

**Issue by 0u812** [sys-bio/roadrunner#126]
Via Matthias:
This one is a killer and made me a lot of headaches (example event_timepoint.py).

The problem is that there are two **identical timepoints** with **different values** associated with a triggered event, one assigning the value before the event, one after the event. Occuring under variable timesteps.
This results in mayor problems in the working with the timecourses afterwards, namely that one timepoint has multiple very different numerical values, which breaks things like interpolations. Many post processing of timecourses rely on uniqueness of timepoints.
If you look at the table at time=10 the PP_S concentrations are 0 and 1 at the same time identical to the time when the event is triggered. The timepoint immediately before the event has to be 10-eps, with eps= min(absTol, relTol*10) and only if there is no timepoint within eps before the event time. The second timepoint has to be exactly 10.

['time', 'PP_S', 'S3', 'J0']
           time, PP_S,       S3,  J0
 [[           0,    0,      0.5,   0],
  [ 0.000686452,    0,      0.5,   0],
  [      6.8652,    0,      0.5,   0],
  [          10,    0,      0.5,   0],
  [          10,    1,      0.5, 0.5],
  [          10,    1, 0.500048, 0.5],
  [     10.4777,    1, 0.977674, 0.5],
  [          15,    1,      5.5, 0.5],
  [          15,    0,      5.5,   0],
  [     15.0008,    0,      5.5,   0],
  [     22.8961,    0,      5.5,   0]]

**Comment by 0u812** [comment]
event_timepoint.py:
```
# -*- coding: utf-8 -*-
"""

@author: Matthias Koenig
@date: 2015-05-05
"""
from __future__ import print_function
import libantimony
import roadrunner
from roadrunner import SelectionRecord

model_txt = """
    model event_timepoint()
    // Reactions
    J0: $PP_S -> 2 S3; K1 * PP_S;

    // Species initializations:
    var species S3;
    S3 = 0.5;

    K1 = 0.5;    
    
    t_peak = 10;
    t_duration = 5;
    peak_status = 0
   
    // Rule for PP_S
    PP_S := peak_status * 1.0 
    
    // Event to trigger the peak
    E0: at(time>=0):  peak_status=0;  
    E1: at(time>=t_peak):  peak_status=1;  
    E2: at(time>=t_peak+t_duration):  peak_status=0;

    end
"""
model = libantimony.loadString(model_txt)
sbml_file = 'event_timepoint.xml'
libantimony.writeSBMLFile(sbml_file, 'event_timepoint')

r = roadrunner.RoadRunner(sbml_file)
print(r.getSBML())
r.selections = ['time'] + r.model.getBoundarySpeciesIds() \
                        + r.model.getFloatingSpeciesIds() \
                        + r.model.getReactionIds()
print(r.selections)

# tolerances & integration
absTol = 1E-12 *min(r.model.getCompartmentVolumes())
relTol = 1E-12

# variable step size integration (reset just to be sure)
r.reset()
r.reset(SelectionRecord.ALL)
r.reset(SelectionRecord.INITIAL_GLOBAL_PARAMETER )
s = r.simulate(0, 20, absolute=absTol, relative=relTol, variableStep=True, stiff=True, plot=True)   

print(s)
```

**Comment by 0u812** [comment]
event_timepoint.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<!-- Created by libAntimony version v2.7.0 on 2015-05-21 15:27 with libSBML version 5.11.4. -->
<sbml xmlns="http://www.sbml.org/sbml/level3/version1/core" level="3" version="1">
  <model id="event_timepoint" name="event_timepoint">
    <listOfCompartments>
      <compartment sboTerm="SBO:0000410" id="default_compartment" spatialDimensions="3" size="1" constant="true"/>
    </listOfCompartments>
    <listOfSpecies>
      <species id="PP_S" compartment="default_compartment" hasOnlySubstanceUnits="false" boundaryCondition="true" constant="false"/>
      <species id="S3" compartment="default_compartment" initialConcentration="0.5" hasOnlySubstanceUnits="false" boundaryCondition="false" constant="false"/>
    </listOfSpecies>
    <listOfParameters>
      <parameter id="K1" value="0.5" constant="true"/>
      <parameter id="t_peak" value="10" constant="true"/>
      <parameter id="t_duration" value="5" constant="true"/>
      <parameter id="peak_status" value="0" constant="false"/>
    </listOfParameters>
    <listOfRules>
      <assignmentRule variable="PP_S">
        <math xmlns="http://www.w3.org/1998/Math/MathML">
          <apply>
            <times/>
            <ci> peak_status </ci>
            <cn type="integer"> 1 </cn>
          </apply>
        </math>
      </assignmentRule>
    </listOfRules>
    <listOfReactions>
      <reaction id="J0" reversible="true" fast="false">
        <listOfReactants>
          <speciesReference species="PP_S" stoichiometry="1" constant="true"/>
        </listOfReactants>
        <listOfProducts>
          <speciesReference species="S3" stoichiometry="2" constant="true"/>
        </listOfProducts>
        <kineticLaw>
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <times/>
              <ci> K1 </ci>
              <ci> PP_S </ci>
            </apply>
          </math>
        </kineticLaw>
      </reaction>
    </listOfReactions>
    <listOfEvents>
      <event id="E0" useValuesFromTriggerTime="true">
        <trigger initialValue="true" persistent="true">
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <geq/>
              <csymbol encoding="text" definitionURL="http://www.sbml.org/sbml/symbols/time"> time </csymbol>
              <cn type="integer"> 0 </cn>
            </apply>
          </math>
        </trigger>
        <listOfEventAssignments>
          <eventAssignment variable="peak_status">
            <math xmlns="http://www.w3.org/1998/Math/MathML">
              <cn type="integer"> 0 </cn>
            </math>
          </eventAssignment>
        </listOfEventAssignments>
      </event>
      <event id="E1" useValuesFromTriggerTime="true">
        <trigger initialValue="true" persistent="true">
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <geq/>
              <csymbol encoding="text" definitionURL="http://www.sbml.org/sbml/symbols/time"> time </csymbol>
              <ci> t_peak </ci>
            </apply>
          </math>
        </trigger>
        <listOfEventAssignments>
          <eventAssignment variable="peak_status">
            <math xmlns="http://www.w3.org/1998/Math/MathML">
              <cn type="integer"> 1 </cn>
            </math>
          </eventAssignment>
        </listOfEventAssignments>
      </event>
      <event id="E2" useValuesFromTriggerTime="true">
        <trigger initialValue="true" persistent="true">
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <geq/>
              <csymbol encoding="text" definitionURL="http://www.sbml.org/sbml/symbols/time"> time </csymbol>
              <apply>
                <plus/>
                <ci> t_peak </ci>
                <ci> t_duration </ci>
              </apply>
            </apply>
          </math>
        </trigger>
        <listOfEventAssignments>
          <eventAssignment variable="peak_status">
            <math xmlns="http://www.w3.org/1998/Math/MathML">
              <cn type="integer"> 0 </cn>
            </math>
          </eventAssignment>
        </listOfEventAssignments>
      </event>
    </listOfEvents>
  </model>
</sbml>
```

**Comment by hsauro** [comment]
Cut down version

from __future__ import print_function
import libantimony
import roadrunner
from roadrunner import SelectionRecord

model_txt = """
    model event_timepoint()
    // Reactions
    J0: $PP_S -> 2 S3; K1 * PP_S;

    // Species initializations:
    var species S3;
    S3 = 0.5;

    K1 = 0.5;    

    t_peak = 10;
    t_duration = 5;

    // Event to trigger the peak
    E1: at(time>=t_peak):  PP_S = 1;
    E2: at(time>=t_peak+t_duration):  PP_S = 0; 

    end
"""
model = libantimony.loadString(model_txt)
sbml_file = 'event_timepoint.xml'
libantimony.writeSBMLFile(sbml_file, 'event_timepoint')

r = roadrunner.RoadRunner(sbml_file)
print(r.getSBML())
r.selections = ['time'] + r.model.getBoundarySpeciesIds() \
                        + r.model.getFloatingSpeciesIds() \
                        + r.model.getReactionIds()
print(r.selections)

# tolerances & integration
absTol = 1E-12 *min(r.model.getCompartmentVolumes())
relTol = 1E-12

# variable step size integration (reset just to be sure)
r.reset()
r.reset(SelectionRecord.ALL)
r.reset(SelectionRecord.INITIAL_GLOBAL_PARAMETER )
s = r.simulate(0, 20, absolute=absTol, relative=relTol, variableStep=True, stiff=True, plot=True)   

print(s)

**Comment by 0u812** [comment]
SBML:
```
<?xml version="1.0" encoding="UTF-8"?>
<!-- Created by libAntimony version v2.7.0 on 2015-06-09 09:53 with libSBML version 5.11.3. -->
<sbml xmlns="http://www.sbml.org/sbml/level3/version1/core" level="3" version="1">
  <model id="event_timepoint" name="event_timepoint">
    <listOfCompartments>
      <compartment sboTerm="SBO:0000410" id="default_compartment" spatialDimensions="3" size="1" constant="true"/>
    </listOfCompartments>
    <listOfSpecies>
      <species id="PP_S" compartment="default_compartment" hasOnlySubstanceUnits="false" boundaryCondition="true" constant="false"/>
      <species id="S3" compartment="default_compartment" initialConcentration="0.5" hasOnlySubstanceUnits="false" boundaryCondition="false" constant="false"/>
    </listOfSpecies>
    <listOfParameters>
      <parameter id="K1" value="0.5" constant="true"/>
      <parameter id="t_peak" value="10" constant="true"/>
      <parameter id="t_duration" value="5" constant="true"/>
    </listOfParameters>
    <listOfReactions>
      <reaction id="J0" reversible="true" fast="false">
        <listOfReactants>
          <speciesReference species="PP_S" stoichiometry="1" constant="true"/>
        </listOfReactants>
        <listOfProducts>
          <speciesReference species="S3" stoichiometry="2" constant="true"/>
        </listOfProducts>
        <kineticLaw>
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <times/>
              <ci> K1 </ci>
              <ci> PP_S </ci>
            </apply>
          </math>
        </kineticLaw>
      </reaction>
    </listOfReactions>
    <listOfEvents>
      <event id="E1" useValuesFromTriggerTime="true">
        <trigger initialValue="true" persistent="true">
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <geq/>
              <csymbol encoding="text" definitionURL="http://www.sbml.org/sbml/symbols/time"> time </csymbol>
              <ci> t_peak </ci>
            </apply>
          </math>
        </trigger>
        <listOfEventAssignments>
          <eventAssignment variable="PP_S">
            <math xmlns="http://www.w3.org/1998/Math/MathML">
              <cn type="integer"> 1 </cn>
            </math>
          </eventAssignment>
        </listOfEventAssignments>
      </event>
      <event id="E2" useValuesFromTriggerTime="true">
        <trigger initialValue="true" persistent="true">
          <math xmlns="http://www.w3.org/1998/Math/MathML">
            <apply>
              <geq/>
              <csymbol encoding="text" definitionURL="http://www.sbml.org/sbml/symbols/time"> time </csymbol>
              <apply>
                <plus/>
                <ci> t_peak </ci>
                <ci> t_duration </ci>
              </apply>
            </apply>
          </math>
        </trigger>
        <listOfEventAssignments>
          <eventAssignment variable="PP_S">
            <math xmlns="http://www.w3.org/1998/Math/MathML">
              <cn type="integer"> 0 </cn>
            </math>
          </eventAssignment>
        </listOfEventAssignments>
      </event>
    </listOfEvents>
  </model>
</sbml>
```

**Comment by 0u812** [comment]
There are actually several issues here:
1. The time values are not unique
2. Matthias requested specific behavior based on his expectation of what the time values should be

(1) is easy to fix. I can make the integrator increment the time value by some epsilon right after event firing. (2) is more difficult. To see why, observe that an event trigger can be an arbitrary Boolean expression. The integrator logic finds the points where these expressions change from false to true, but, like all numerical solutions, the value is approximate. If I were to tackle (2), I would implement a solution that backtracks to the point where the trigger evaluates to false - that would be the "pre-trigger" timepoint. However, event triggers do not necessarily depend purely on time, so a full solution would require reverse-time integration.

In any case, we won't be able to guarantee that the two time values are exactly 10-eps and 10 resp. like Matthias requested. All we will be able to guarantee is that the event trigger evaluates to false for the pre-trigger timepoint and true for the post-trigger timepoint - neither will necessarily be exactly 10. That's if we fix (2). If we just fix (1), then the timepoints will merely be unique.

**Comment by luciansmith** [comment]
Ack!  No, you absolutely shouldn't increment the time value after event firing.  That is an incorrect interpretation of the model. As far as I can tell, nothing is wrong about the integrator.  The problem is that extra output is being written out that should not be written out.

The issue is that the simulator is being given a model and a bunch of time points to write out.  Those time points are fixed expected output:  if the modeler requested 11 points between 0 and 10 inclusive, the output should have exactly 11 lines, corresponding to time points 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, and 10.  If the model experienced an event triggering at one of those time points, or multiple events triggering at one of those time points, you cannot write out 12 points or 15 points or anything:  you *must* write out exactly the 11 points.

Similarly, you want the *last* value that all model variables had written out.  This is why incrementing the time step by epsilon absolutely will not work.  If an event triggers at time point 3, and you increment time incorrectly, the new value will not be noted as having taken place at time point 3, but at time point 3.0000001, and hence won't be written out.  Instead, the values that the model used to have at time point 3 will be written out, which will be incorrect.

The whole issue with working backward to figure out when an event triggered is a solved problem, and hence, a red herring.  The integrator is already behaving correctly.  What is incorrect is the way it is writing out the values.

**Comment by 0u812** [comment]
@luciansmith not sure if I understand. This is a variable time step simulation, so the time values are not explicitly requested (so you don't know if you'll end up with 11 points or whatever). In Matthias' model, he had an event that triggered at time 10 s, and none of the reported values were before 10 s (i.e. there was no timepoint for 9.9999...). So, our pre-trigger timepoint (where PP_S = 0) had the wrong time value - it was after 10 s instead of before it. The backtracking idea I mentioned was to fix that part of the problem, i.e. make the output:
```
[t], [PP_S]
9.9999, 0
10.00001, 1
```

However, currently we have something like:
```
[t], [PP_S]
10.00001, 0
10.00001, 1
```
(the numerical values are slightly greater than 10 if you dump out the digits)
The idea I suggested of incrementing the time value would be to make these unique:
```
[t], [PP_S]
10.00001, 0
10.00002, 1
```
Could you clarify your comment?

**Comment by matthiaskoenig** [comment]
Hi all,
the two important points for me are
1) no duplicate timepoints
2) at least:
a) one time point exactly when the event is triggered:the numerical values
of the model for this timepoint are the values after evaluating the event,
i.e. all the event assignments.
b) one time point as short as possible before the event is triggered with
the numerical values corresponding to before the event was evaluated.
The things with the tolerances was just a proposal where to put this second
time point. I am not picky about when to exactly put it, but it should be
so close before the event that the change of numerical values not affected
by the event assignments are below the tolerances.

The time point b) is extremely important otherwise all plots look
completely wrong (linear increase from the last integration timepoint to
the event time point, with the last integration timepoint often very far
away from the actual event.

Matt

On Wed, Jun 10, 2015 at 8:50 PM, luciansmith <notifications@github.com>
wrote:

> Ack! No, you absolutely shouldn't increment the time value after event
> firing. That is an incorrect interpretation of the model. As far as I can
> tell, nothing is wrong about the integrator. The problem is that extra
> output is being written out that should not be written out.
>
> The issue is that the simulator is being given a model and a bunch of time
> points to write out. Those time points are fixed expected output: if the
> modeler requested 11 points between 0 and 10 inclusive, the output should
> have exactly 11 lines, corresponding to time points 0, 1, 2, 3, 4, 5, 6, 7,
> 8, 9, and 10. If the model experienced an event triggering at one of those
> time points, or multiple events triggering at one of those time points, you
> cannot write out 12 points or 15 points or anything: you *must* write out
> exactly the 11 points.
>
> Similarly, you want the *last* value that all model variables had written
> out. This is why incrementing the time step by epsilon absolutely will not
> work. If an event triggers at time point 3, and you increment time
> incorrectly, the new value will not be noted as having taken place at time
> point 3, but at time point 3.0000001, and hence won't be written out.
> Instead, the values that the model used to have at time point 3 will be
> written out, which will be incorrect.
>
> The whole issue with working backward to figure out when an event
> triggered is a solved problem, and hence, a red herring. The integrator is
> already behaving correctly. What is incorrect is the way it is writing out
> the values.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sys-bio/roadrunner/issues/126#issuecomment-110875547>.
>



-- 
---------------------------------------------------
Matthias König
Computational Systems Biochemistry
Institute of Biochemistry
Charité - Universitätsmedizin Berlin
http://www.charite.de/sysbio/people/koenig/
Tel: + 49 30 450 528 197
Tel: + 49 176 81168480
---------------------------------------------------

**Comment by 0u812** [comment]
Thanks for your input, @matthiaskoenig! I think I have a solution to at least (1), but let me work with Lucian to see if this is the right way to fix it. Once we have a fix I can give you a Linux build to try it out. We have an Apt repository with beta .deb packages for roadrunner now.

**Comment by matthiaskoenig** [comment]
@Kyle thanks so much
I already managed to build from source with your instructions, but will
give the apt-repository a try when the fix is commited.



On Wed, Jun 10, 2015 at 9:28 PM, Kyle Medley <notifications@github.com>
wrote:

> Thanks for your input, @matthiaskoenig <https://github.com/matthiaskoenig>!
> I think I have a solution to at least (1), but let me work with Lucian to
> see if this is the right way to fix it. Once we have a fix I can give you a
> Linux build to try it out. We have an Apt repository with beta .deb
> packages for roadrunner now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sys-bio/roadrunner/issues/126#issuecomment-110885868>.
>



-- 
---------------------------------------------------
Matthias König
Computational Systems Biochemistry
Institute of Biochemistry
Charité - Universitätsmedizin Berlin
http://www.charite.de/sysbio/people/koenig/
Tel: + 49 30 450 528 197
Tel: + 49 176 81168480
---------------------------------------------------

**Comment by luciansmith** [comment]
OK, flexible time points does indeed make this trickier, particularly if you want a time point output *before* the current time point.

Here's my suggested algorithm:
* Whenever an event triggers and its delay has been carried out (but before assignment) assume that at time point (current minus delta), the values were the same. Output results from that time point.
* Now calculate your events.  Multiple events might fire; if so, don't output anything until the last event has fired at that time point.
* After time moves forward again, output the final model state at that time point.

Note that this will not catch changes that happen 'during' a single time point, but merely the state of the model before the time point, and the state of the model after the time point has passed.  Imagine the following model:

x = 0
y = 0

at(time>=5): x=2;
at(x>1): p=0, y=5

this would output
time        x   y
0             0  0
4.99999  0  0
5             0  5

In other words, it wouldn't catch that x changed from 0 to 2 and back to 0, all at time point 5, but it would catch the side-effect of x being greater than 1 at that point, namely, y becoming 5.

**Comment by 0u812** [comment]
Thanks Lucian! You bring up a good point about multiple events firing at the same time point. I'll ask Herbert what he thinks about that situation when he gets back.

On June 10, 2015 1:44:32 PM PDT, luciansmith <notifications@github.com> wrote:
>OK, flexible time points does indeed make this trickier, particularly
>if you want a time point output *before* the current time point.
>
>Here's my suggested algorithm:
>* Whenever an event triggers and its delay has been carried out (but
>before assignment) assume that at time point (current minus delta), the
>values were the same. Output results from that time point.
>* Now calculate your events.  Multiple events might fire; if so, don't
>output anything until the last event has fired at that time point.
>* After time moves forward again, output the final model state at that
>time point.
>
>Note that this will not catch changes that happen 'during' a single
>time point, but merely the state of the model before the time point,
>and the state of the model after the time point has passed.  Imagine
>the following model:
>
>x = 0
>y = 0
>
>at(time>=5): x=2;
>at(x>1): p=0, y=5
>
>this would output
>time        x   y
>0             0  0
>4.99999  0  0
>5             0  5
>
>In other words, it wouldn't catch that x changed from 0 to 2 and back
>to 0, all at time point 5, but it would catch the side-effect of x
>being greater than 1 at that point, namely, y becoming 5.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/sys-bio/roadrunner/issues/126#issuecomment-110907135

-- 
Sent from my TARDIS with K-9 Mail. Wibbly-wobbly wordy gurdy.

**Comment by 0u812** [comment]
I sent Matthias the fixed binaries in a separate email. The fix is in the timepoints branch. I'm going to leave this issue open to remind me to merge it in.

**Comment by 0u812** [comment]
Fixed in #158
